lib/CL/pocl_cache.c

```diff
@@ -0,0 +1,470 @@
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "config.h"
+#include "kernellib_hash.h"
+
+#include "pocl_hash.h"
+#include "pocl_cache.h"
+#include "pocl_file_util.h"
+
+#include "pocl_cl.h"
+#include "pocl_runtime_config.h"
+
+#define POCL_LAST_ACCESSED_FILENAME "/last_accessed"
+/* The filename in which the program's build log is stored */
+#define POCL_BUILDLOG_FILENAME      "/build.log"
+/* The filename in which the program source is stored in the program's temp dir. */
+#define POCL_PROGRAM_CL_FILENAME "/program.cl"
+/* The filename in which the program LLVM bc is stored in the program's temp dir. */
+#define POCL_PROGRAM_BC_FILENAME "/program.bc"
+
+static char cache_topdir[POCL_FILENAME_LENGTH];
+static int cache_topdir_initialized = 0;
+
+int pocl_cl_device_to_index(cl_program   program,
+                            cl_device_id device) {
+    unsigned i;
+    assert(program);
+    for (i = 0; i < program->num_devices; i++)
+        if (program->devices[i] == device)
+            return i;
+    return -1;
+}
+
+static void program_device_dir(char*        path,
+                              cl_program   program,
+                              unsigned     device_i,
+                              char*        append_path) {
+    assert(path);
+    assert(program);
+    assert(device_i < program->num_devices);
+    /* sanity check on SHA1 digest emptiness */
+    assert(program->build_hash[device_i][0] > 0);
+
+    int bytes_written = snprintf(path, POCL_FILENAME_LENGTH,
+                                 "%s/%s%s", cache_topdir,
+                                 program->build_hash[device_i],
+                                 append_path);
+    assert(bytes_written > 0 && bytes_written < POCL_FILENAME_LENGTH);
+}
+
+
+// required in llvm API
+void pocl_cache_program_bc_path(char*        program_bc_path,
+                                cl_program   program,
+                                unsigned     device_i) {
+    program_device_dir(program_bc_path, program,
+                       device_i, POCL_PROGRAM_BC_FILENAME);
+}
+
+// required in llvm API
+void pocl_cache_kernel_so_path(char* kernel_so_path, cl_program program,
```

Comment:
It's actually "work_group_function_so_path", there can be many per kernel, one for each local size.